proto/buf/alpha/registry/v1alpha1/doc.proto

```diff
@@ -377,6 +378,47 @@ message FieldOptions {
   int32 jstype = 4;
 }
 
+// FieldLiteral represents a field and its value. It can be used to model descriptor options.
+message FieldLiteral {
```

Comment:
I don't actually see this used anywhere, where it is used?

Also how was the API created/designed?

Reply:
The idea is to supplement the fields that look like `FieldOptions field_options = 10` (across all of the descriptor types) with a field that instead looks like `repeated FieldLiteral options = 11`. This representation can model any option and is specifically intended for modeling custom options in a way that the frontend can render into the docs.

It's not used anywhere yet at [my suggestion](https://github.com/bufbuild/buf/pull/2752#pullrequestreview-1863908719). As it is, this can be pulled into the other repo and the transformation from the `protosource.OptionExtensionDescriptor` model to the `[]*FieldLiteral` model can be implemented and tested there w/out (yet) impacting the BSR API.

But I'm not super-familiar with the typical process of implementing BSR APIs. Is it common to update the API first and then leave this stuff in the API, but unimplemented, for however long it takes to actually implement? It seemed better to me to try to implement the server-side stuff alongside the API change and have them reviewed simultaneously (or concurrently?), to eliminate useless churn (or introducing compatibility issues) in the APIs if implementation reveals that changes are required.

Reply:
I'm mostly interested how the specific messages were designed - are these net-new, or were they pulled from somewhere? If new-new, I need to review them more carefully. If not net-new, where did they derive from?

Reply:
These are net new

Reply:
Using `FieldOptions` (or other option types) makes consuming the value much more complicated since it will have extensions. So a consumer would need to have the schemata for all custom options in order to interpret the data.

Basically same goes for well-known types -- the closest would be `google.protobuf.Value`, but it represents JSON data (so does not distinguish between any of the numeric types, does not support enums except as string or int values, and supports null which isn't appropriate here). Aside from that, the only thing that can dynamically represent any kind of message is `google.protobuf.Any` and that has the same drawbacks as extensions: the consumer must have the schemata for everything in order to interpret this.

This is _kind of like_ `google.protobuf.Value`, except it is for modeling arbitrary Protobuf values instead of JSON values. It's intended to provide a very easy way to work with the data without the schema.

Reply:
While we could use the `google.protobuf.*Options` messages here, we would also have to ship the descriptors for the extensions and types they use. This only makes sense to me if we use descriptors for the entire package docs, replacing the current layer. This is feasible and has advantages, but it requires extra steps to split descriptor sets by package... and would obviously be a major pivot. 

`google.protobuf.Value` cannot store the values NaN and Infinity for `DOUBLE`. It's unlikely that they appear in option values, but they can. It also cannot represent `BYTES`. This means that the data will contain strings with special meaning - base64 encoded `BYTES`, `"NaN"`, `"Infinity"`, and `"-Infinity"` - that the client cannot distinguish from other strings without knowing the schema, which leads to the same problem as with `google.protobuf.*Options`.

I agree with Josh here.